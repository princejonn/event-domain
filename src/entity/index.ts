export * from "./Aggregate";
export * from "./Saga";
export * from "./View";
