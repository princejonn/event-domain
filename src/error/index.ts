export * from "./AggregateAlreadyCreatedError";
export * from "./AggregateDestroyedError";
export * from "./AggregateNotCreatedError";
export * from "./AggregateNotDestroyedError";
export * from "./CausationMissingEventsError";
export * from "./CommandSchemaValidationError";
export * from "./ConcurrencyError";
export * from "./DomainError";
export * from "./EventStreamRevisionError";
export * from "./HandlerNotRegisteredError";
export * from "./IllegalEntityChangeError";
export * from "./MessageTypeError";
export * from "./MongoDuplicateKeyError";
export * from "./MongoNotUpdatedError";
export * from "./SagaAlreadyCreatedError";
export * from "./SagaCausationError";
export * from "./SagaDestroyedError";
export * from "./SagaNotCreatedError";
export * from "./SagaNotInStoreError";
export * from "./SagaRevisionError";
export * from "./ViewAlreadyCreatedError";
export * from "./ViewCausationError";
export * from "./ViewDestroyedError";
export * from "./ViewNotCreatedError";
export * from "./ViewRevisionError";
