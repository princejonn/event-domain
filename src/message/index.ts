export * from "./Command";
export * from "./DomainEvent";
export * from "./Message";
export * from "./Timeout";
