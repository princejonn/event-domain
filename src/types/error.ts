import { LindormErrorOptions } from "@lindorm-io/errors";

export interface DomainErrorOptions extends LindormErrorOptions {
  permanent: boolean;
}
