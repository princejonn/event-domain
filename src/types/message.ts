import { AggregateIdentifier } from "./aggregate";
import { MessageType } from "../enum";

export interface IMessage {
  aggregate: AggregateIdentifier;
  causationId: string;
  correlationId: string;
  data: Record<string, any>;
  delay: number;
  id: string;
  mandatory: boolean;
  name: string;
  timestamp: Date;
  type: MessageType;
}

export interface MessageOptions {
  aggregate: AggregateIdentifier;
  causationId?: string;
  correlationId?: string;
  data?: Record<string, any>;
  delay?: number;
  id?: string;
  mandatory?: boolean;
  name: string;
  timestamp?: Date | string;
  type?: MessageType;
}
