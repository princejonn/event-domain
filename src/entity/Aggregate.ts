import { AGGREGATE_MERGE_STATE_SCHEMA, getApplyEventSchema } from "../schema";
import { AggregateEventHandler } from "../handler";
import { Command, DomainEvent } from "../message";
import { Logger } from "@lindorm-io/winston";
import { assertCamelCase, validateSchemaAsync, validateSchemaSync } from "../util";
import { cloneDeep, find, merge, set } from "lodash";
import {
  AggregateData,
  AggregateEventHandlerContext,
  AggregateIdentifier,
  AggregateOptions,
  AnyType,
  GenericRecord,
  IAggregate,
} from "../types";
import {
  AggregateDestroyedError,
  AggregateNotDestroyedError,
  HandlerNotRegisteredError,
  IllegalEntityChangeError,
  MessageTypeError,
} from "../error";

export class Aggregate<State> implements IAggregate<State> {
  readonly id: string;
  readonly name: string;
  readonly context: string;

  private readonly _eventHandlers: Array<AggregateEventHandler<State>>;
  private readonly _events: Array<DomainEvent>;
  private readonly _state: GenericRecord<State>;
  private _destroyed: boolean;
  private _destroying: boolean;
  private _numberOfLoadedEvents: number;

  private readonly logger: Logger;

  constructor(options: AggregateOptions<State>, logger: Logger) {
    this.logger = logger.createChildLogger(["Aggregate"]);

    assertCamelCase(options.name);

    this.id = options.id;
    this.name = options.name;
    this.context = options.context;

    this._destroyed = false;
    this._destroying = false;
    this._eventHandlers = options.eventHandlers || [];
    this._events = [];
    this._numberOfLoadedEvents = 0;
    this._state = {};
  }

  public get destroyed(): boolean {
    return this._destroyed;
  }
  public set destroyed(_) {
    throw new IllegalEntityChangeError();
  }

  public get events(): Array<DomainEvent> {
    return this._events;
  }
  public set events(_) {
    throw new IllegalEntityChangeError();
  }

  public get numberOfLoadedEvents(): number {
    return this._numberOfLoadedEvents;
  }
  public set numberOfLoadedEvents(_) {
    throw new IllegalEntityChangeError();
  }

  public get state(): GenericRecord<State> {
    return this._state;
  }
  public set state(_) {
    throw new IllegalEntityChangeError();
  }

  public async apply(
    causation: Command,
    name: string,
    data?: Record<string, any>,
  ): Promise<void> {
    this.logger.debug("applying command", { causation, name, data });

    await this.handleEvent(
      new DomainEvent(
        {
          aggregate: { id: this.id, name: this.name, context: this.context },
          data: data,
          name: name,
        },
        causation,
      ),
    );
  }

  public async load(event: DomainEvent): Promise<void> {
    this.logger.debug("loading event", { event });

    await this.handleEvent(event);
    this._numberOfLoadedEvents += 1;
  }

  private async handleEvent(event: DomainEvent): Promise<void> {
    this.logger.debug("handling event", { event });

    if (!(event instanceof DomainEvent)) {
      throw new MessageTypeError(event, DomainEvent);
    }

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    const destroying = this._destroying;

    await validateSchemaAsync(
      getApplyEventSchema({
        id: this.id,
        name: this.name,
        context: this.context,
      }),
      {
        name: event.name,
        aggregate: event.aggregate,
        data: event.data,
        causationId: event.causationId,
      },
    );

    const eventHandler: AggregateEventHandler<State> = find(
      this._eventHandlers,
      (handler) => handler.eventName === event.name,
    );

    if (!(eventHandler instanceof AggregateEventHandler)) {
      throw new HandlerNotRegisteredError();
    }

    const context: AggregateEventHandlerContext<State> = {
      event: event,
      state: cloneDeep(this._state),
      logger: this.logger.createChildLogger(["AggregateEventHandler"]),
      destroy: this.destroy.bind(this),
      destroyNext: this.destroyNext.bind(this),
      mergeState: this.mergeState.bind(this),
      setState: this.setState.bind(this),
    };

    await eventHandler.handler(context);

    if (destroying && !this._destroyed) {
      throw new AggregateNotDestroyedError();
    }

    this._events.push(event);

    this.logger.debug("successfully handled event", { event });
  }

  private destroy(): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._destroyed = true;
    this._destroying = false;
  }

  private destroyNext(): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._destroying = true;
  }

  private mergeState(data: Record<string, any>): void {
    validateSchemaSync(AGGREGATE_MERGE_STATE_SCHEMA, { data });

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    merge(this._state, data);
  }

  private setState(path: string, value: AnyType): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    set(this._state, path, value);
  }

  public toJSON(): AggregateData<State> {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      destroyed: this.destroyed,
      events: cloneDeep(this.events),
      numberOfLoadedEvents: cloneDeep(this.numberOfLoadedEvents),
      state: cloneDeep(this.state),
    };
  }

  static identifier(identifier: AggregateIdentifier): string {
    return `${identifier.context}.${identifier.name}.${identifier.id}`;
  }
}
