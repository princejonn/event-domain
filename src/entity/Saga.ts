import { Command, DomainEvent, Timeout } from "../message";
import {
  ISaga,
  SagaIdentifier,
  SagaOptions,
  AnyType,
  GenericRecord,
  SagaData,
} from "../types";
import { IllegalEntityChangeError, MessageTypeError, SagaDestroyedError } from "../error";
import { SAGA_DISPATCH_SCHEMA, SAGA_UPDATE_STATE_SCHEMA } from "../schema";
import { assertCamelCase, validateSchemaSync } from "../util";
import { cloneDeep, merge, set } from "lodash";
import { Logger } from "@lindorm-io/winston";

export class Saga<State> implements ISaga<State> {
  readonly id: string;
  readonly name: string;
  readonly context: string;

  private readonly _causationList: Array<string>;
  private readonly _messagesToDispatch: Array<Command | DomainEvent>;
  private readonly _revision: number;
  private readonly _state: GenericRecord<State>;
  private _destroyed: boolean;

  private readonly logger: Logger;

  constructor(options: SagaOptions<State>, logger: Logger) {
    this.logger = logger.createChildLogger(["Saga"]);

    assertCamelCase(options.name);

    this.id = options.id;
    this.name = options.name;
    this.context = options.context;

    this._causationList = options.causationList || [];
    this._destroyed = options.destroyed || false;
    this._messagesToDispatch = options.messagesToDispatch || [];
    this._revision = options.revision || 0;
    this._state = options.state || {};
  }

  public get causationList(): Array<string> {
    return this._causationList;
  }
  public set causationList(_) {
    throw new IllegalEntityChangeError();
  }

  public get messagesToDispatch(): Array<Command> {
    return this._messagesToDispatch;
  }
  public set messagesToDispatch(_) {
    throw new IllegalEntityChangeError();
  }

  public get destroyed(): boolean {
    return this._destroyed;
  }
  public set destroyed(_) {
    throw new IllegalEntityChangeError();
  }

  public get revision(): number {
    return this._revision;
  }
  public set revision(_) {
    throw new IllegalEntityChangeError();
  }

  public get state(): GenericRecord<State> {
    return this._state;
  }
  public set state(_) {
    throw new IllegalEntityChangeError();
  }

  public destroy(): void {
    this.logger.debug("destroying");

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._destroyed = true;
  }

  public dispatch(causation: DomainEvent, command: Command): void {
    this.logger.debug("dispatching command", { causation, command });

    validateSchemaSync(SAGA_DISPATCH_SCHEMA, { command, causation });

    if (!(command instanceof Command)) {
      throw new MessageTypeError(command, Command);
    }

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._messagesToDispatch.push(
      new Command(
        merge(
          {
            aggregate: {
              context: this.context,
            },
          },
          command,
        ),
        causation,
      ),
    );
  }

  public mergeState(data: Record<string, any>): void {
    this.logger.debug("merging state", { data });

    validateSchemaSync(SAGA_UPDATE_STATE_SCHEMA, { data });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    merge(this._state, data);
  }

  public setState(path: string, value: AnyType): void {
    this.logger.debug("setting state", { path, value });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    set(this._state, path, value);
  }

  public timeout(
    causation: DomainEvent,
    name: string,
    data: Record<string, any>,
    delay: number,
  ): void {
    this.logger.debug("dispatching timeout", { causation, name, data, delay });

    if (this._destroyed) {
      throw new SagaDestroyedError();
    }

    this._messagesToDispatch.push(
      new Timeout(
        {
          name,
          data,
          aggregate: {
            id: this.id,
            name: this.name,
            context: this.context,
          },
          delay,
        },
        causation,
      ),
    );
  }

  public toJSON(): SagaData<State> {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      causationList: cloneDeep(this.causationList),
      messagesToDispatch: cloneDeep(this.messagesToDispatch),
      destroyed: this.destroyed,
      revision: this.revision,
      state: cloneDeep(this.state),
    };
  }

  static identifier(identifier: SagaIdentifier): string {
    return `${identifier.context}.${identifier.name}.${identifier.id}`;
  }
}
