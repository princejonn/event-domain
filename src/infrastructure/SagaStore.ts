import {
  SagaIdentifier,
  ISagaStore,
  SagaStoreOptions,
  SagaStoreSaveOptions,
} from "../types";
import { Logger } from "@lindorm-io/winston";
import { Message } from "../message";
import { Saga } from "../entity";
import { SagaStoreInMemory } from "./in-memory";
import { SagaStoreMongo } from "./mongo";
import { StoreType } from "../enum";
import { CLEAR_SAGA_MESSAGES_TO_DISPATCH_SCHEMA, SAVE_SAGA_SCHEMA } from "../schema";
import { validateSchemaAsync } from "../util";

export class SagaStore<State> {
  private readonly logger: Logger;
  private sagaStore: ISagaStore<State>;

  constructor(logger: Logger, options?: SagaStoreOptions) {
    this.logger = logger.createChildLogger(["SagaStore"]);

    const type = options?.type || StoreType.MEMORY;

    switch (type) {
      case StoreType.CUSTOM:
        this.sagaStore = options.custom;
        break;

      case StoreType.MEMORY:
        this.sagaStore = new SagaStoreInMemory(this.logger);
        break;

      case StoreType.MONGO:
        this.sagaStore = new SagaStoreMongo(this.logger, options.mongo);
        break;

      default:
        throw new Error("Invalid Store Type");
    }
  }

  public async save(
    saga: Saga<State>,
    causation: Message,
    options?: SagaStoreSaveOptions,
  ): Promise<Saga<State>> {
    await validateSchemaAsync(SAVE_SAGA_SCHEMA, {
      saga,
      causation,
      messagesToDispatch: saga.messagesToDispatch,
    });

    try {
      const result = await this.sagaStore.save(saga, causation, options);

      this.logger.debug("successfully saved saga", { saga: Saga.identifier(saga) });
      return result;
    } catch (err) {
      this.logger.error("failed to save saga", { saga: Saga.identifier(saga) });
      throw err;
    }
  }

  public async load(sagaIdentifier: SagaIdentifier): Promise<Saga<State>> {
    try {
      const result = await this.sagaStore.load(sagaIdentifier);

      if (!result) {
        this.logger.debug("successfully created saga", {
          saga: Saga.identifier(sagaIdentifier),
        });
        return new Saga({ ...sagaIdentifier, messagesToDispatch: [] }, this.logger);
      }

      this.logger.debug("successfully loaded saga", { saga: Saga.identifier(result) });
      return result;
    } catch (err) {
      this.logger.error("failed to load saga", { saga: Saga.identifier(sagaIdentifier) });
      throw err;
    }
  }

  public async clearMessagesToDispatch(saga: Saga<State>): Promise<Saga<State>> {
    await validateSchemaAsync(CLEAR_SAGA_MESSAGES_TO_DISPATCH_SCHEMA, {
      saga,
      messagesToDispatch: saga.messagesToDispatch,
    });

    try {
      const result = await this.sagaStore.clearMessagesToDispatch(saga);

      this.logger.debug("successfully cleared messages to dispatch", {
        saga: Saga.identifier(saga),
      });

      return result;
    } catch (err) {
      this.logger.error("failed to clear messages to dispatch", {
        saga: Saga.identifier(saga),
      });

      throw err;
    }
  }
}
