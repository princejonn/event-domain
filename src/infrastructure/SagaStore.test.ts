import MockDate from "mockdate";
import { Command, DomainEvent } from "../message";
import { Saga } from "../entity";
import { SagaStore } from "./SagaStore";
import { logger } from "../test";
import { SagaCausationError, SagaNotInStoreError, SagaRevisionError } from "../error";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("SagaStore", () => {
  const aggregateIdentifier = {
    id: "2a553a74-0f84-4a7d-b91e-2902666db98c",
    name: "aggregateName",
    context: "aggregateContext",
  };
  const sagaIdentifier = {
    id: "2a553a74-0f84-4a7d-b91e-2902666db98c",
    name: "sagaName",
    context: "sagaContext",
  };

  let sagaStore: SagaStore<any>;
  let event: DomainEvent;

  beforeEach(() => {
    sagaStore = new SagaStore(logger);

    event = new DomainEvent(
      {
        id: "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
        name: "eventName",
        aggregate: aggregateIdentifier,
        data: { eventData: true },
      },
      new Command({
        id: "86e9a7ee-7455-4438-9c61-7e41bd0191d5",
        name: "commandName",
        aggregate: aggregateIdentifier,
      }),
    );
  });

  describe("constructor", () => {
    test("should create a saga store", () => {
      expect(new SagaStore(logger)).toStrictEqual(expect.any(SagaStore));
    });

    test("should throw if saga store type is erroneous", () => {
      expect(
        () =>
          new SagaStore({
            // @ts-ignore
            type: "wrong",
          }),
      ).toThrow();
    });
  });

  describe("save", () => {
    test("should save a saga", async () => {
      await expect(
        sagaStore.save(
          new Saga(
            {
              ...sagaIdentifier,
              messagesToDispatch: [],
            },
            logger,
          ),
          event,
        ),
      ).resolves.toMatchSnapshot();
    });

    test("should throw if causation already exists in list", async () => {
      await expect(
        sagaStore.save(
          new Saga(
            {
              ...sagaIdentifier,
              causationList: [event.id],
              messagesToDispatch: [],
            },
            logger,
          ),
          event,
        ),
      ).rejects.toThrow(SagaCausationError);
    });

    test("should throw if revision is wrong", async () => {
      const saga = await sagaStore.save(
        new Saga(
          {
            ...sagaIdentifier,
            causationList: [],
            messagesToDispatch: [],
          },
          logger,
        ),
        event,
      );

      const event2 = new DomainEvent(
        {
          id: "55b4e3ef-3a9c-4f72-9c6a-765ed59bc1b0",
          name: "eventName2",
          aggregate: aggregateIdentifier,
          data: { eventData: true },
        },
        new Command({
          id: "a38165e1-1336-45cf-a428-377231ed9af1",
          name: "commandName2",
          aggregate: aggregateIdentifier,
        }),
      );

      await expect(
        sagaStore.save(
          new Saga(
            {
              ...saga,
              id: saga.id,
              name: saga.name,
              context: saga.context,
              causationList: saga.causationList,
              messagesToDispatch: [],
              destroyed: saga.destroyed,
              revision: 99,
              state: saga.state,
            },
            logger,
          ),
          event2,
        ),
      ).rejects.toThrow(SagaRevisionError);
    });
  });

  describe("load", () => {
    test("should load an existing saga", async () => {
      await sagaStore.save(
        new Saga(
          {
            ...sagaIdentifier,
            messagesToDispatch: [],
            state: { exists: true },
          },
          logger,
        ),
        event,
      );

      await expect(sagaStore.load(sagaIdentifier)).resolves.toMatchSnapshot();
    });

    test("should load a new saga", async () => {
      await expect(sagaStore.load(sagaIdentifier)).resolves.toMatchSnapshot();
    });
  });

  describe("clearMessagesToDispatch", () => {
    test("should successfully clear list of commands", async () => {
      const command2 = new Command({
        id: "cd4bcb6b-ce9d-4cc3-93a0-bc57434c0af6",
        name: "command2",
        aggregate: aggregateIdentifier,
      });

      const saga = await sagaStore.save(
        new Saga(
          {
            ...sagaIdentifier,
            messagesToDispatch: [command2],
          },
          logger,
        ),
        event,
      );

      await expect(sagaStore.clearMessagesToDispatch(saga)).resolves.toMatchSnapshot();
    });

    test("should throw if saga does not exist in store", async () => {
      const saga = new Saga(
        {
          ...sagaIdentifier,
          messagesToDispatch: [],
        },
        logger,
      );

      await expect(sagaStore.clearMessagesToDispatch(saga)).rejects.toThrow(
        SagaNotInStoreError,
      );
    });
  });
});
