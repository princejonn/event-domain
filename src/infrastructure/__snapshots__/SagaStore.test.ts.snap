// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SagaStore clearMessagesToDispatch should successfully clear list of commands 1`] = `
Object {
  "causationList": Array [
    "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
  ],
  "context": "sagaContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "messagesToDispatch": Array [],
  "name": "sagaName",
  "revision": 2,
  "state": Object {},
}
`;

exports[`SagaStore load should load a new saga 1`] = `
Object {
  "causationList": Array [],
  "context": "sagaContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "messagesToDispatch": Array [],
  "name": "sagaName",
  "revision": 0,
  "state": Object {},
}
`;

exports[`SagaStore load should load an existing saga 1`] = `
Object {
  "causationList": Array [
    "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
  ],
  "context": "sagaContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "messagesToDispatch": Array [],
  "name": "sagaName",
  "revision": 1,
  "state": Object {
    "exists": true,
  },
}
`;

exports[`SagaStore save should save a saga 1`] = `
Object {
  "causationList": Array [
    "6d35d053-fc89-4ee5-93d6-87c0ceecb8b3",
  ],
  "context": "sagaContext",
  "destroyed": false,
  "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
  "messagesToDispatch": Array [],
  "name": "sagaName",
  "revision": 1,
  "state": Object {},
}
`;
